### PR TITLE
Use pure rust kmean in IVF and PQ

### PR DESCRIPTION
**Pull request by eddyxu**
Support python SDK to call `dataset.create_index()`

### CODE REVIEW

File: rust/src/index/vector/pq.rs [hunk 1]
```diff
@@ -229,12 +230,15 @@ impl ProductQuantizer {
...
+        let rng = rand::rngs::SmallRng::from_entropy();
+
         // TODO: parallel training.
         for sub_vec in &sub_vectors {
             // Centroids for one sub vector.
             let values = sub_vec.values();
             let flatten_array: &Float32Array = as_primitive_array(&values);
-            let centroids = train_kmeans(flatten_array, dimension, num_centorids, 100)?;
+            let centroids =
+                train_kmeans(flatten_array, dimension, num_centorids, 25, rng.clone()).await?;
```
Comment (changhiskhan) [inline rust/src/index/vector/pq.rs:241]: Why change to 25?

File: rust/src/index/vector/kmeans.rs [hunk 2]
```diff
@@ -15,43 +15,71 @@
...
+        for idx in chosen.iter() {
+            let s = array.slice(idx * dimension, dimension);
+            builder.append_slice(as_primitive_array::<Float32Type>(s.as_ref()).values());
+        }
+        builder.finish()
+    } else {
+        array.clone()
+    };
+
+    println!(
```
Comment (changhiskhan) [inline rust/src/index/vector/kmeans.rs:77]: Print?

**Review by changhiskhan — APPROVED** [review]

File: rust/src/index/vector/pq.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/index/vector/pq.rs:241]: this was the default value in the faiss kmean we used. seems to get good enough results.

File: rust/src/index/vector/kmeans.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/index/vector/kmeans.rs:77]: fixed.

**Pull request merged**